# dbscATAC: a resource of single-cell super-enhancers/enhancers and gene markers for scATAC-seq data
![headnew](https://github.com/user-attachments/assets/a1ae902b-3291-4f3b-a069-0b04f8eeaf7a)
dbscATAC is a specialized database offering comprehensive annotations of 213,835 super-enhancers across 520 tissue/cell types from 3 species, as well as 347,484 gene markers, 13,470,526 enhancers, and 10,402,346 enhancer-gene interactions, and 43,202,785 TF binding sites derived from 1,668,076 single cells of scATAC-seq data spanning 1,028 tissue/cell types in 13 species. The predicted results are available in an interactive website http://singlecelldb.com/dbscATAC/index.php.

# Overview of dbscATAC
![Fig1](https://github.com/user-attachments/assets/184c25ed-988c-4671-9fdc-3a6be4d8c92b)<br />

# Dependencies
To conduct prediction of super-enhancers, typical enhancers, and gene markers, following softwares are required:<br />
✯ ROSE<br />
✯ R version > 4.2<br />
✯ Signac 1.12.0<br />
✯ Seurat 5.1.0<br />
✯ cicero 1.3.9<br />
✯ Perl >5.16.3<br />
✯ bedtools > 2<br />
✯ macs2 2.2.6<br />
✯ sinto 0.10.1<br />
✯ deeptools > 2<br />
✯ python3<br />

# Tutorial
**[Prediction-of-superenhancers](https://github.com/EvansGao/dbscATAC/wiki/Prediction-of-superenhancers)**   : This will use GSE for example

  
# Download of results
All the single-cell chromatin accessibility annotation data are available at:<br />
http://singlecelldb.com/dbscATAC/download.php
